### DISCUSSION TITLE
Deploy action only appears to work when committing to master branch

### DISCUSSION THREAD

**Discussion by jamesmrollins** [JamesIves/github-pages-deploy-action#493] [Q&A] [answered]
Hi All - I have a rather low skill-level in this arena, but have spent several days experimenting and looking for answers on my own.  So, thanks in advance for you help. 

**Application:**  I am rendering a bookdown manual from markdown files using executables from r-lib.  Then these html files are deposited to a gh-pages branch to provision a website.  The actions supporting this are located in the master branch and triggered when there is a push to Master.   In this version (code below), the rendering and deployment processes work, when someone edits a source .md file in the gh-pages branch and then commits another, unrelated file directly to Master.

**Code**

     on:
       push:
          branches: [ master ] 

     name: renderbook

     jobs:
      bookdown:
         name: Render-Book
         runs-on: macOS-latest
    steps:
      - uses: actions/checkout@v2
        with:
          ref: gh-pages
      - uses: r-lib/actions/setup-r@v1
      - uses: r-lib/actions/setup-pandoc@v1
      - name: Install rmarkdown
        run: Rscript -e 'install.packages(c("rmarkdown","bookdown","HelpersMG","readr"))'
        
      - uses: r-lib/actions/setup-tinytex@v1
      - name : install tinytex
        run: tlmgr --version
    
      - name: Render Book
        run: |
             Rscript -e 'bookdown::render_book("index.Rmd", "bookdown::epub_book", output_dir="docs")'
             Rscript -e 'bookdown::render_book("index.Rmd", "bookdown::word_document2", output_dir="docs")'
             Rscript -e 'bookdown::render_book("index.Rmd", "bookdown::gitbook", output_dir="docs")'
      - uses: actions/upload-artifact@v1
        with:
          name: docs
          path: docs/
     # deploy to gh pages yml
       build-and-deploy:
        runs-on: ubuntu-latest
        needs: bookdown
        steps:
           - name: Checkout 🛎️
             uses: actions/checkout@v2 # If you're using actions/checkout@v2 you must set persist-credentials to false in most cases for the deployment to work correctly.
        with:
               persist-credentials: false
        
           - name: Download Artifacts 🔻 # The built project is downloaded into the 'site' folder.
             uses: actions/download-artifact@v1
             with:
               name: docs
          
           - name: Deploy 🚀
             uses: JamesIves/github-pages-deploy-action@releases/v3
             with:
               GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
               BRANCH: gh-pages # The branch the action should deploy to.
               FOLDER: docs
               TARGET_FOLDER: docs # The folder the action should deploy.
               CLEAN: true

 **PROBLEM:**  We are trying to deploy a branch merge routine where a feature branch is created, files are edited, and committed back to the feature (copy) branch, and the render/deploy actions trigger with a pull request is opened to merge the feature branch with the gh-pages branch.  When this code below is implemented in both the gh-pages and the copy feature branch, the action will render, but it doesn't deploy the files to the /doc file.  Below is a screenshot of the deploy action report.

I have tried various permutations creating and merging duplicates away from the gh-pages branch, but nothing seems to work except actions triggered on Master as shown above.

**Code**

    on:
     pull_request:
     branches: [ gh-pages ]

     name: renderbook

     jobs:
       bookdown:
         name: Render-Book
         runs-on: macOS-latest
         steps:
           - uses: actions/checkout@v2
             with:
               ref: gh-pages
           - uses: r-lib/actions/setup-r@v1
           - uses: r-lib/actions/setup-pandoc@v1
           - name: Install rmarkdown
             run: Rscript -e 'install.packages(c("rmarkdown","bookdown","HelpersMG","readr"))'
        
           - uses: r-lib/actions/setup-tinytex@v1
           - name : install tinytex
             run: tlmgr --version
         
           - name: Render Book
             run: |
                  Rscript -e 'bookdown::render_book("index.Rmd", "bookdown::epub_book", output_dir="docs")'
                  Rscript -e 'bookdown::render_book("index.Rmd", "bookdown::word_document2", output_dir="docs")'
                  Rscript -e 'bookdown::render_book("index.Rmd", "bookdown::gitbook", output_dir="docs")'
           - uses: actions/upload-artifact@v1
             with:
               name: docs
               path: docs/
     # deploy to gh pages yml
       build-and-deploy:
        runs-on: ubuntu-latest
        needs: bookdown
        steps:
           - name: Checkout 🛎️
             uses: actions/checkout@v2 # If you're using actions/checkout@v2 you must set persist-credentials to false in most cases for the deployment to work correctly.
             with:
             persist-credentials: false
        
          - name: Download Artifacts 🔻 # The built project is downloaded into the 'site' folder.
            uses: actions/download-artifact@v1
            with:
            name: docs
          
           - name: Deploy 🚀
             uses: JamesIves/github-pages-deploy-action@releases/v3
             with:
               GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
               BRANCH: gh-pages # The branch the action should deploy to.
               FOLDER: docs
               TARGET_FOLDER: docs # The folder the action should deploy.
               CLEAN: true

**Screenshot**
When the deploy doesn't work,  there is a report on line 54 where the deploy "exits early."  This report doesn't happen when triggered from the Master branch.  

![image](https://user-images.githubusercontent.com/30132017/98162984-fa685000-1e96-11eb-8207-c8c41aefbc22.png)

**Comment by JamesIves** [comment, marked as answer]
It looks like you're modifying the tree before the commits are made and it's reversing the timeline. Have you tried setting `PRESERVE` to `true` in your workflow? In a future version this option will be toggled on by default.

**Comment by jamesmrollins** [comment]
Hi @JamesIves  - Thanks a million for your help!  Your suggestion worked perfectly - thank you.
